cpp/arcticdb/storage/rocksdb/rocksdb_storage.cpp

```diff
@@ -0,0 +1,239 @@
+
+    auto lib_dir = root_path / lib_path_str;
+    auto db_name = lib_dir.generic_string();
+
+    std::set<std::string> key_names;
+    arcticdb::entity::foreach_key_type([&](KeyType&& key_type) {
+        std::string handle_name = fmt::format("{}", key_type);
+        key_names.insert(handle_name);
+    });
+
+    std::vector<::rocksdb::ColumnFamilyDescriptor> column_families;
+    ::rocksdb::DBOptions db_options;
+
+    ::rocksdb::ConfigOptions cfg_opts; // not used
+    auto s = ::rocksdb::LoadLatestOptions(cfg_opts, db_name, &db_options, &column_families);
+    if (s.ok()) {
+        std::set<std::string> existing_key_names{};
+        for (const auto& desc : column_families) {
+            if (desc.name != ::rocksdb::kDefaultColumnFamilyName) {
+                existing_key_names.insert(desc.name);
+            }
+        }
+        util::check(existing_key_names == key_names, "Existing database has incorrect key columns.");
+    } else if (s.IsNotFound()) {
+        util::check_arg(mode > OpenMode::READ, "Missing dir {} for lib={}. mode={}",
+                            db_name, lib_path_str, mode);
+        // Default column family required, error if not provided.
+        column_families.emplace_back(::rocksdb::kDefaultColumnFamilyName, ::rocksdb::ColumnFamilyOptions());
+        for (const auto& key_name: key_names) {
+            util::check(key_name != ::rocksdb::kDefaultColumnFamilyName,
+                        "Key name clash with mandatory default column name");
+            column_families.emplace_back(key_name, ::rocksdb::ColumnFamilyOptions());
+        }
+        fs::create_directories(lib_dir);
+        db_options.create_if_missing = true;
+        db_options.create_missing_column_families = true;
+    } else {
+        util::raise_rte(DEFAULT_ROCKSDB_NOT_OK_ERROR + s.ToString());
+    }
+
+    std::vector<::rocksdb::ColumnFamilyHandle*> handles;
+    // Note: the "default" handle will be returned as well. It is necessary to delete this, but not
+    // rocksdb's internal handle to the default column family as returned by DefaultColumnFamily().
+    s = ::rocksdb::DB::Open(db_options, db_name, column_families, &handles, &db_);
+    util::check(s.ok(), DEFAULT_ROCKSDB_NOT_OK_ERROR + s.ToString());
+    util::check(handles.size() == column_families.size(), "Open returned wrong number of handles.");
+    for (std::size_t i = 0; i < handles.size(); i++) {
+        handles_by_key_type_.emplace(column_families[i].name, handles[i]);
+    }
+    handles.clear();
+}
+
+RocksDBStorage::~RocksDBStorage() {
+    for (const auto& [key_type_name, handle]: handles_by_key_type_) {
+        auto s = db_->DestroyColumnFamilyHandle(handle);
+        util::check(s.ok(), DEFAULT_ROCKSDB_NOT_OK_ERROR + s.ToString());
+    }
+    handles_by_key_type_.clear();
+    auto s = db_->Close();
+    util::check(s.ok(), DEFAULT_ROCKSDB_NOT_OK_ERROR + s.ToString());
+    delete db_;
```

Comment:
Because the column handles need to be deleted before `db_`, we could still put `db_` in a `std::unique_ptr` but I don't see any particular value in this.

As I explained in this PR's description, we can't make column handles and the `db_` all be `std::unique_ptr` because then we couldn't easily control destruction order without being very careful about the order which these attributes are defined in the class.

One other idea would be to create an object to take care of the connection. Maybe something like `RocksConnection` which just stores the `db_` and the column family handles. Then this little class could handle the messy destructor logic, and therefore isolate it out of the `RocksDBStorage` class. I'm not sure there's much added value in doing this.

One thing to note though is that a custom destructor for `RocksDBStorage`, as defined here, does invalidate the default move constructor which is actually required for when this class is used in `storage_factory.cpp` as it is written at the moment. In my next PR though I will be re-factoring `Storage` objects to delete all move and copy constructors (they shouldn't be move or copy constructible!), and therefore re-write the anti-pattern used in `storage_factory.cpp`. This PR's code is therefore slightly broken if we actually were to attempt to move construct a `RocksDBStorage.cpp` object. There is no need to make the changes here though.